docs/proposals/L7-protocols.md

```diff
@@ -0,0 +1,190 @@
+ * Tag `protocol` is optional
+ * If tag `protocol` is missing, `TCP` will be assumed (current behaviour)
+
+E.g.,
+
+```yaml
+type: Dataplane
+mesh: default
+name: backend
+networking:
+  inbound:
+  - interface: 192.168.0.1:18080:8080
+    tags:
+      service: backend
+      protocol: http   # notice `protocol` tag
+```
+
+#### Kubernetes mode
+
+* Since `Dataplane` resource is auto-generated on `k8s`, a user should edit other `k8s` resources instead
+* To define application protocol explictly, a user should edit `Service` resource (rather than `Deployment` / `Pod`)
+  * The reason for that is that `k8s` does not require `Deployment` / `Pod` to explicitly mention all ports exposed by the app
+  * On the other hand, `k8s` does require `Service` to explicitly mention all its ports
+* Annotations might be too verbose and redundant if a user already sticks to minimal conventions when naming service ports
+  * E.g., if a user already gives to service ports names like `http`, `http-metrics`, `grpc`, `postgres`, that information is (optimistically) enough for `Kuma`
+* If a user prefers explicit annotations, e.g. to make it very clear why this particular piece of configuration is important, he should use `<port>.service.kuma.io/protocol`
+ * In case if there is no explicit annotation and port naming conventions are not followed, `TCP` protocol will be assumed (current behaviour)
+
+E.g.,
+
+* A user might opt for minimal port naming conventions (`protocol` can appear anywhere inside the port `name`)
+
+  ```yaml
+  apiVersion: v1
+  kind: Service
+  metadata:
+    name: backend
+    namespace: example
+  spec:
+    ports:
+    - port: 8080
+      name: http          # `http` protocol in the `name`
+    - port: 7070
+      name: http-metrics  # `http` protocol in the `name`
+    - port: 6060
+      name: somethinggrpc # `grpc` protocol in the `name`
+    selector:
+      app: backend
+  ```
+
+* A user might opt for explicit annotations in the form `<port>.service.kuma.io/protocol`
+
+  ```yaml
+  apiVersion: v1
+  kind: Service
+  metadata:
+    name: backend
+    namespace: example
+    annotations:
+      8080.service.kuma.io/protocol: http # explicit `http` protocol
+      7070.service.kuma.io/protocol: http # explicit `http` protocol
+      6060.service.kuma.io/protocol: grpc # explicit `grpc` protocol
+  spec:
+    ports:
+    - port: 8080
+      name: a
+    - port: 7070
+      name: b
+    - port: 6060
+      name: tcp  # `protocol` from annotation will take precedence
+    selector:
+      app: backend
+  ```
+
+### Configuration model #2
+
+* User describes a high-level service, `Kuma` infers low-level technical aspects, such as `protocol`
+
+#### Universal mode
+
+* REST API
+  ```yaml
+  type: ServiceProfile
+  mesh: default
+  name: backend # the same value as in `service` tag
+  profiles:
+  - api:
+      rest:
+        spec:
+          url: https://generator.swagger.io/api/swagger.json
```

Comment:
What can we do with such information?
LinkerD has this to generate endpoints for which then metrics are collected, but we don't need it once generate listener with http connection manager filter.